### PR TITLE
Avoid false positive warning about using `sync_dist` when using torchmetrics

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

See title. Reported on slack.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -525,7 +525,7 @@ def _get_cache(result_metric: _ResultMetric, on_step: bool) -> Optional[Tensor]:
         elif not on_step and result_metric.meta.on_epoch:
             if result_metric._computed is None:
                 should = result_metric.meta.sync.should
-                if not result_metric.meta.sync.should and distributed_available():
+                if not result_metric.meta.sync.should and distributed_available() and result_metric.is_tensor:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/logger_connector/result.py:528]:
```suggestion
                if not should and distributed_available() and result_metric.is_tensor:
```

File: tests/tests_pytorch/core/test_metric_result_integration.py [hunk 2]
```diff
@@ -666,19 +668,27 @@ def on_train_start(self):
...
+
+    # self.log('bar', 0.5, ..., sync_dist=False)
     meta = _Metadata("foo", "bar")
     meta.sync = _Sync(_should=False)
-    result_metric = _ResultMetric(metadata=meta, is_tensor=True)
-    result_metric.update(torch.tensor(7.0), 10)
+    is_tensor = isinstance(log_val, torch.Tensor)
+
+    if not is_tensor:
+        log_val.update(torch.Tensor([0, 1]), torch.LongTensor([0, 0]))
```
Comment (carmocca) [inline tests/tests_pytorch/core/test_metric_result_integration.py:681]: The class constructors are not recommended

```suggestion
        log_val.update(torch.tensor([0, 1]), torch.tensor([0, 0], dtype=torch.long))
```

**Review by carmocca — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
